<launch>

  <!-- Command Line Arguments for gscam -->
  <arg name="sensor_id" default="0" />                       <!-- The sensor id of the camera -->
  <arg name="cam_name" default="csi_cam_$(arg sensor_id)" /> <!-- The name of the camera (corrsponding to the camera info) -->
  <arg name="frame_id" default="/$(arg cam_name)_link" />    <!-- The TF frame ID. -->
  <arg name="sync_sink" default="true" />                    <!-- Synchronize the app sink. Setting this to false may resolve problems with sub-par framerates. -->
  <arg name="width" default="1280" />                        <!-- Image Width -->
  <arg name="height" default="720" />                       <!-- Image Height -->
  <arg name="fps" default="21" />                            <!-- Desired framerate. True framerate may not reach this if set too high. -->

  <!-- This guy isn't working for some reason ... -->
  <!-- For Including Camera calibration data (using camera_calibration package) -->
  <!-- <arg name="cam_cal_data" default="$(find urVision)/config/$(arg cam_name).yaml" /> -->

  <!-- Make arguments available to parameter server -->
  <param name="$(arg cam_name)/camera_id" type="int" value="$(arg sensor_id)" />
  <param name="$(arg cam_name)/image_width" type="int" value="$(arg width)" />
  <param name="$(arg cam_name)/image_height" type="int" value="$(arg height)" />
  <param name="$(arg cam_name)/target_fps" type="int" value="$(arg fps)" />

  <!-- Define the GSCAM pipeline -->
  <env name="GSCAM_CONFIG" value="nvarguscamerasrc ! video/x-raw(memory:NVMM),
    width=(int)$(arg width), height=(int)$(arg height), format=(string)NV12, framerate=(fraction)$(arg fps)/1 ! 
    nvvidconv flip-method=2 ! video/x-raw, format=(string)BGRx ! videoconvert ! video/x-raw, format=(string)BGR" />

  <!-- Start the GSCAM node -->
  <node pkg="gscam" type="gscam" name="$(arg cam_name)">
    <param name="camera_name" value="$(arg cam_name)" />
    <param name="frame_id" value="$(arg frame_id)" />
    <param name="sync_sink" value="$(arg sync_sink)" />
    <!-- This guy isn't working for some reason ... -->
    <!-- <param name="camera_info_url" value="$(arg cam_cal_data)" /> -->
    <remap from="camera/image_raw" to="$(arg cam_name)/image_raw" />
    <remap from="/set_camera_info" to="$(arg cam_name)/set_camera_info" />
  </node>

    <!-- Provide rectification -->
  <node pkg="image_proc" type="image_proc" name="creative_image_proc"
        ns="$(arg cam_name)" />

  <arg name="startup_msg" default="CSI Camera Stream Started Successfully!" />
    <node name="pub_text" pkg="rostopic" type="rostopic"
    args="pub /startup_msg std_msgs/String '$(arg startup_msg)'"
    output="screen" />
    <node name="print_text" pkg="rostopic" type="rostopic"
    args="echo /startup_msg" output="screen" />
        
</launch>
